azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
         return agent_family_manifests[0]
 
     @staticmethod
-    def __get_requested_version(agent_family):
+    def _get_version_from_gs(agent_family):
         """
-        Get the requested version from agent family
-        Returns: Requested version if supported and available in the GS
-                 None if requested version missing or GA versioning not enabled
+        Get the version from agent family
+        Returns: version if supported and available in the GS
+                 None if version is missing
         """
-        if conf.get_enable_ga_versioning() and agent_family.is_requested_version_specified:
-            if agent_family.requested_version is not None:
-                return FlexibleVersion(agent_family.requested_version)
+        if agent_family.version is not None:
+            return FlexibleVersion(agent_family.version)
         return None
 
-    @staticmethod
-    def __get_largest_version(agent_manifest):
-        largest_version = FlexibleVersion("0.0.0.0")
-        for pkg in agent_manifest.pkg_list.versions:
-            pkg_version = FlexibleVersion(pkg.version)
-            if pkg_version > largest_version:
-                largest_version = pkg_version
-        return largest_version
+    def run(self, goal_state):
+
+        try:
+            # Ignore new agents if update is disabled. The latter flag only used in e2e tests.
+            if not self._autoupdate_enabled or not conf.get_download_new_agents():
+                return
+
```

Comment:
As discussed previously, lines 218 to 229 are executed in the main loop every 6 seconds. Those lines invoke methods than in turn invoke other methods. That is too much logic. It's way too easy to introduce bugs that produce too much telemetry, too much log output, consume too much CPU, etc. Please work on moving the check for the timer currently at line 232 right after the check for update enable on line 215.